### DISCUSSION TITLE
About the `Session token`, `Access token` and `proxy` used for login and authentication by this project.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [acheong08/ChatGPT#886] [General]
Based on the description [here](https://github.com/acheong08/ChatGPT#authentication-method-choose-1), I've the following puzzles:

1. Where can I find my `session_token`?
2. For the `Access token`, as explained [here](https://blog.laoda.de/archives/chatgpt-and-openai), there are two types of api keys: `OPENAI API KEY` and `ChatGPT API KEY`. I'm not sure which type should I used for this tool. And furthermore, it seems that the [Access token](https://chat.openai.com/api/auth/session) has a very different form from the API keys. What is the difference between them, and what exactly should I use with this tool?
3. Which type of `proxy` is supported by tool?

Any tips will be appreciated.

Regards,
Zhao

**Comment by linhandev** [comment]
For q3 I think you can refer to this https://www.python-httpx.org/advanced/#http-proxying. I've tried socks5 link works.

**Comment by acheong08** [comment]
https://github.com/acheong08/ChatGPT/wiki/Authentication

**Reply by acheong08** [reply]
As for proxy, you must specify the protocol: `socks5://IP:PORT` or `http://IP:PORT`. Note that authentication isn't fully supported. You might run into bugs there.

**Reply by hongyi-zhao** [reply]
> As for proxy, you must specify the protocol: `socks5://IP:PORT` or `http://IP:PORT`. 

What about `socks5h`?

> Note that authentication isn't fully supported. You might run into bugs there.

Do you mean the `user:pass` part, if used, is not guaranteed to work in the proxy?

**Reply by acheong08** [reply]
> What about socks5h?

dunno, never tried

> Do you mean the user:pass part, if used, is not guaranteed to work in the proxy?

Yes. I've encountered bugs with authenticated proxies in the past. It might not be reliable

**Reply by linhandev** [reply]
> https://github.com/acheong08/ChatGPT/wiki/Authentication

It seems OpenAIAuth.Authenticator can retrieve access_token after logging in with email&pswd. Currently the access_token is only recorded in the request header. If it's also set to Chatbot.config['access_token'], I think user can record this value after logging in, save it in file and use it for the next login. I think this would saves the trouble of logging in with browser. @acheong08

**Reply by acheong08** [reply]
>  I think this would saves the trouble of logging in with browser

OpenAIAuth is a reverse engineering of Auth0 so that it does not require a browser

**Reply by linhandev** [reply]
> > I think this would saves the trouble of logging in with browser
> 
> OpenAIAuth is a reverse engineering of Auth0 so that it does not require a browser

Yeah, I meant OpenAIAuth got the access_token without a browser but the access_token isn't exposed to user in the Chatbot. If it is accessible after logging in at Chatbot.config['access_token'], users can save it and use the value at next startup programmatically.

**Reply by acheong08** [reply]
I suppose saving the access_token would be a nice feature. There is a mechanism from within the access_token to check for expiry without contacting OpenAI as well.

Will accept pull request or I'll add tomorrow

**Reply by hongyi-zhao** [reply]
> https://github.com/acheong08/ChatGPT/wiki/Authentication

According to [the following description](https://github.com/acheong08/ChatGPT/wiki/Authentication#methods-of-authentication) given on the above wiki page:

> You cannot put multiple forms of authentication into a config at the same time

So, one and only one type of them is needed and enough for this purpose, and [the last method](https://github.com/acheong08/ChatGPT/wiki/Authentication#access-token) is the recommended one, based on the following statement on the above-mentioned wiki page:

> Recommended method of authentication

**Reply by acheong08** [reply]
> So, one and only one type of them is needed and enough for this purpose, and [the last method](https://github.com/acheong08/ChatGPT/wiki/Authentication#access-token) is the recommended one, based on the following statement on the above-mentioned wiki page:

Yes. It is the most stable method that has pretty much no bugs. You'll have to update it every two weeks though. I'll add some code to allow the input of multiple methods of authentication tomorrow.

**Reply by linhandev** [reply]
> There is a mechanism from within the access_token to check for expiry without contacting OpenAI as well.

I think in the worse case we can try to send out a testing message with a random content and watch out for some sort of access_token invalid error response.

> Will accept pull request or I'll add tomorrow

It's probably better if you do it. I'm not familiar with the code base and would probably introduce bugs. :rofl:

**Reply by hongyi-zhao** [reply]
Please refer the following comment [here](https://blog.laoda.de/archives/chatgpt-and-openai):

![image](https://user-images.githubusercontent.com/11155854/221338092-10aca941-c97d-4e44-9147-06f3e1ec1c38.png)

I've simply translated the above description as follows, FYI:

> If you have used the ChatGPT's official chat and OpenAI's API interface chat, respectively, you will find that the API interface chat is much more "stupid" than the official ChatGPT chat.

So, I'd like to know which of these types of interfaces, or both, are being called in the background by this project?

**Reply by acheong08** [reply]
> So, I'd like to know which of these types of interfaces, or both, are being called in the background by this project?

They're the exact same as the one on `chat.openai.com`. Dunno why it returns worse content. Maybe they got a plus account and forgot to set `paid` to true in the config

**Reply by hongyi-zhao** [reply]
Thank you for your comment.

Another question: As far as code development and assistance are concerned, what's your point of view of [GitHub Copilot](https://github.com/features/copilot) vs ChatGPT?

**Reply by acheong08** [reply]
Depends on the language. I would be comfortable using GitHub Copilot in Go due to its strictness which means that it's less prone to AI error. As for Python, I try to avoid both because it's a bit too difficult to spot errors while simply reading the code.

**Reply by acheong08** [reply]
I've pretty much never used ChatGPT for copy pasting code but rather just suggesting libraries to use. I do have Copilot installed though disabled by default

**Reply by hongyi-zhao** [reply]
If I understand correctly, it seems that [GitHub Copilot](https://github.com/features/copilot) doesn't provide a free plan at all, while [ChatGPT does](https://openai.com/api/pricing/).

**Reply by acheong08** [reply]
Students get it for free

**Reply by hongyi-zhao** [reply]
What is the free quota?

**Reply by acheong08** [reply]
> What is the free quota?

No quota
